### PR TITLE
feat: mailto backend

### PR DESCRIPTION
**Pull request by Kneesal**
# Description

### <samp>🤖 Generated by Copilot at a03c642</samp>

This pull request adds the email action feature to the `apps/api-journeys` app, which allows users to send emails to visitors who interact with journeys. It introduces new GraphQL types, inputs, mutations, and fields to the schema, and implements the resolver and service logic for the email action. It also updates the existing button event module to handle the email action as a possible button action. It adds the `lastEmailAction` field to the visitor module to track the last email action that a visitor received. It includes unit tests for the new and modified classes and methods.

-[ Link to Basecamp Todo](https://3.basecamp.com/3105655/buckets/32387559/todos/6170464030)

# How should this PR be QA Tested?

Please describe the QA tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

# Walkthrough

### <samp>🤖 Generated by Copilot at a03c642</samp>

*  Add a new feature to allow users to send emails to visitors who interact with journeys
  * Define a new enum value `EmailAction` for the `ActionType` enum to represent a new type of action that can be triggered by a button click event ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-e60078ed0fc71b8cd3bb2fca25505e7dba866b7d463147662a12432dfe287273R53), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-bef781d45ccbb6775960a589c940a77584bd143e501a8ae9b83d699d803561f2L138-R139))
  * Define a new type `EmailAction` and a new input type `EmailActionInput` to specify the shape and structure of the email action data that can be queried and mutated ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-e60078ed0fc71b8cd3bb2fca25505e7dba866b7d463147662a12432dfe287273R351-R366), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-3cbd0d77b8eff95fb0f62db7f9f1a4b69ccd568ae3902a3450bac7d752c817a3R58-R63), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-3cbd0d77b8eff95fb0f62db7f9f1a4b69ccd568ae3902a3450bac7d752c817a3R84-R88), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-bef781d45ccbb6775960a589c940a77584bd143e501a8ae9b83d699d803561f2R227-R231), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-bef781d45ccbb6775960a589c940a77584bd143e501a8ae9b83d699d803561f2R635-R641), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-69e724008c5d9f9b438c1ce579d7a4a5d2c92eea451a3c642eaae44fdb481437R34-R39), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-69e724008c5d9f9b438c1ce579d7a4a5d2c92eea451a3c642eaae44fdb481437R60-R64))
  * Add a new mutation field `blockUpdateEmailAction` to the `Mutation` type to provide a way for the frontend to update the email action of a block in a journey ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-e60078ed0fc71b8cd3bb2fca25505e7dba866b7d463147662a12432dfe287273R816), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-3cbd0d77b8eff95fb0f62db7f9f1a4b69ccd568ae3902a3450bac7d752c817a3R95), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-bef781d45ccbb6775960a589c940a77584bd143e501a8ae9b83d699d803561f2R1149-R1150), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-69e724008c5d9f9b438c1ce579d7a4a5d2c92eea451a3c642eaae44fdb481437R87-R91))
  * Add a new field `lastEmailAction` to the `Visitor` type to provide a way for the frontend to display and track the email actions that visitors have taken ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-e60078ed0fc71b8cd3bb2fca25505e7dba866b7d463147662a12432dfe287273R2447-R2452), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-3cbd0d77b8eff95fb0f62db7f9f1a4b69ccd568ae3902a3450bac7d752c817a3R1836-R1841), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-bef781d45ccbb6775960a589c940a77584bd143e501a8ae9b83d699d803561f2R1105), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-d38571c2fa6cfb52ac34c8b29546437aa9d780de4c77fd7dda2ba006233da652R128-R132), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-db278dbb0e00e2c314e0cf51ba0476f446f35b296b863bed1dfcfe03e519e781R38))
  * Add the `EmailAction` type to the union types `Action` and `ButtonAction` to make the email action type a possible member of the action type and the button action type ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-3cbd0d77b8eff95fb0f62db7f9f1a4b69ccd568ae3902a3450bac7d752c817a3R901), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-f9dd7205341a75daad609ca6fb983cbb5169a1bb36a23f79cadcc3d8ec675449R6))
  * Implement the mutation logic for updating the email action of a block in a journey in the `EmailActionResolver` class in the `apps/api-journeys/src/app/modules/action/emailAction/emailAction.resolver.ts` file ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-4bd62b4b4f9247353282dc562aaffb4c252d2a29310a259230ffcbff80b6b6b9R1-R77))
    * Import and register the `EmailActionResolver` class as a provider for the action module in the `apps/api-journeys/src/app/modules/action/action.module.ts` file ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-c25b51134beb7ccbcf61a3c454538a10881a3d34023627f181584516af2497f1R11), [link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-c25b51134beb7ccbcf61a3c454538a10881a3d34023627f181584516af2497f1R21))
    * Write unit tests for the `EmailActionResolver` class in the `apps/api-journeys/src/app/modules/action/emailAction/emailAction.resolver.spec.ts` file ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-928436cdc0e776ae4e4c28ac26c546f37d76adb1895eed3a06c8a5389e1092e1R1-R108))
  * Implement the mutation logic for creating a button click event when a visitor interacts with a button block in a journey in the `ButtonClickEventResolver` class in the `apps/api-journeys/src/app/modules/event/button/button.resolver.ts` file ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-8781cd0196e95d3108a8cfbcecde16f30ff5b9be217a68057022e7766c65ab10R55-R62))
    * Write a unit test case for the `ButtonClickEventResolver` class when the button action is an email action in the `apps/api-journeys/src/app/modules/event/button/button.resolver.spec.ts` file ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-a5bd79c1d24a16f9021044536fa8fa49606cb4fd2829e97c06c643100e164c3eR92-R107))
  * Implement the resolver logic for determining the concrete type of an action based on its fields in the `ActionResolver` class in the `apps/api-journeys/src/app/modules/action/action.resolver.ts` file ([link](https://github.com/JesusFilm/core/pull/1576/files?diff=unified&w=0#diff-63e5dfbf51551571ecf00f17b3ba0141a316f2662a8748f44565cb9c872771ccR23))

### CODE REVIEW

File: apps/api-journeys/db/migrations/20230529025612_20230529025610/migration.sql
Comment (tataihono) [inline apps/api-journeys/db/migrations/20230529025612_20230529025610/migration.sql:1]: this file has gotta go

File: apps/api-journeys/db/migrations/20230529025650_20230529025649/migration.sql
Comment (tataihono) [inline apps/api-journeys/db/migrations/20230529025650_20230529025649/migration.sql:1]: same for this one

File: libs/journeys/ui/src/libs/action/actionFields.ts [hunk 1]
```diff
@@ -19,5 +19,8 @@ export const ACTION_FIELDS = gql`
     ... on LinkAction {
       url
     }
+    ... on EmailAction {
+      email
+    }
```
Comment (tataihono) [inline libs/journeys/ui/src/libs/action/actionFields.ts:24]: is this required?

**Review by tataihono — CHANGES_REQUESTED** [review]

File: libs/journeys/ui/src/libs/action/actionFields.ts (hunk 1 above)
Comment (Kneesal) [reply libs/journeys/ui/src/libs/action/actionFields.ts:24]: Without this field, it generates the wrong types. e.g. 
```
ActionField_NavigateToAction: {
...ActionField_NavigateToAction.props
__typename: NavigateToAction | EmailAction
} 

```

File: apps/api-journeys/db/migrations/20230529025650_20230529025649/migration.sql
Comment (Kneesal) [reply apps/api-journeys/db/migrations/20230529025650_20230529025649/migration.sql:1]: agreed, however:
[https://www.prisma.io/docs/concepts/components/prisma-migrate/migration-histories](https://www.prisma.io/docs/concepts/components/prisma-migrate/migration-histories)

File: apps/api-journeys/db/migrations/20230529025612_20230529025610/migration.sql
Comment (Kneesal) [reply apps/api-journeys/db/migrations/20230529025612_20230529025610/migration.sql:1]: agreed, however:
[https://www.prisma.io/docs/concepts/components/prisma-migrate/migration-histories](https://www.prisma.io/docs/concepts/components/prisma-migrate/migration-histories)

**Pull request merged**